### ISSUE TITLE
passenger-status does not display, troubleshooting issue, Ubuntu 18.04

### ISSUE THREAD

**Issue by pkarjala** [phusion/passenger#2249]
**Question 1: What is the problem?**
- What is the expected behavior?
- What is the actual behavior?
- How can we reproduce it? Please try to provide a sample application (or Virtual Machine) demonstrating the issue. Otherwise, if we can't reproduce it, we might have to ask you a number of followup questions or run certain commands to try and figure out the problem.

Be as detailed as possible in your descriptions, include any logs and stack traces (don't just cut/paste the error, provide some logging before that too).

Your answer:

When running `passenger-status`, it should display the status of the Passenger server.  Instead, it responds with the following message:

```
ERROR: Phusion Passenger doesn't seem to be running. If you are sure that it is running, then the causes of this problem could be:

1. You customized the instance registry directory using Apache's PassengerInstanceRegistryDir option, Nginx's passenger_instance_registry_dir option, or Phusion Passenger Standalone's --instance-registry-dir command line argument. If so, please set the environment variable PASSENGER_INSTANCE_REGISTRY_DIR to that directory and run passenger-status again.
2. The instance directory has been removed by an operating system background service. Please set a different instance registry directory using Apache's PassengerInstanceRegistryDir option, Nginx's passenger_instance_registry_dir option, or Phusion Passenger Standalone's --instance-registry-dir command line argument.
```

I have searched the following threads for resolution:

https://stackoverflow.com/questions/31761542/phusion-passenger-status-what-value-for-passenger-instance-registry-dir
https://github.com/phusion/passenger/issues/821

I have read the documentation for `passenger-status` and `PassengerInstanceRegistryDir` at
https://www.phusionpassenger.com/library/admin/apache/overall_status_report.html 
https://www.phusionpassenger.com/library/config/apache/reference/#passengerinstanceregistrydir 

I am able to run the following without issue:

```
$ sudo passenger-memory-stats
Version: 6.0.4
Date   : 2020-01-28 11:10:20 -1000

---------- Apache processes ----------
PID    PPID   VMSize    Private  Name
--------------------------------------
24843  1      168.2 MB  0.5 MB   /usr/sbin/apache2 -k start
24861  24843  834.8 MB  0.8 MB   /usr/sbin/apache2 -k start
24862  24843  834.8 MB  0.6 MB   /usr/sbin/apache2 -k start
### Processes: 3
### Total private dirty RSS: 1.86 MB


-------- Nginx processes --------

### Processes: 0
### Total private dirty RSS: 0.00 MB


----- Passenger processes -----
PID    VMSize    Private  Name
-------------------------------
24845  390.1 MB  2.5 MB   Passenger watchdog
24849  675.5 MB  4.1 MB   Passenger core
25050  336.2 MB  44.9 MB  Passenger AppPreloader: /var/www/kaiapuni-test (forking...)
25069  336.3 MB  12.9 MB  Passenger AppPreloader: /var/www/kaiapuni-test (forking...)
25090  336.4 MB  12.9 MB  Passenger AppPreloader: /var/www/kaiapuni-test (forking...)
### Processes: 5
### Total private dirty RSS: 77.41 MB
```

```
$ passenger-config validate-install
What would you like to validate?
Use <space> to select.
If the menu doesn't display correctly, press '!'

   ⬢  Passenger itself
 ‣ ⬢  Apache

-------------------------------------------------------------------------

Checking whether there are multiple Apache installations...
Only a single installation detected. This is good.

-------------------------------------------------------------------------

 * Checking whether this Passenger install is in PATH... ✓
 * Checking whether there are no other Passenger installations... ✓
 * Checking whether Apache is installed... ✓
 * Checking whether the Passenger module is correctly configured in Apache... ✓

Everything looks good. :-)
```

```
$ passenger-config about detect-apache2
Looking for possible Apache installations...
 --> Looking for /usr/local/rvm/gems/ruby-2.4.9/bin/apxs2: not found
 --> Looking for /usr/local/rvm/gems/ruby-2.4.9@global/bin/apxs2: not found
 --> Looking for /usr/local/rvm/rubies/ruby-2.4.9/bin/apxs2: not found
 --> Looking for /usr/local/sbin/apxs2: not found
 --> Looking for /usr/local/bin/apxs2: not found
 --> Looking for /usr/sbin/apxs2: not found
 --> Looking for /usr/bin/apxs2: found
 --> Looking for /sbin/apxs2: not found
 --> Looking for /bin/apxs2: not found
 --> Looking for /usr/games/apxs2: not found
 --> Looking for /usr/local/games/apxs2: not found
 --> Looking for /snap/bin/apxs2: not found
 --> Looking for /usr/local/rvm/bin/apxs2: not found
 --> Looking for /home/ubuntu/.rvm/bin/apxs2: not found
 --> Looking for /usr/local/rvm/gems/ruby-2.4.9/bin/apxs: not found
 --> Looking for /usr/local/rvm/gems/ruby-2.4.9@global/bin/apxs: not found
 --> Looking for /usr/local/rvm/rubies/ruby-2.4.9/bin/apxs: not found
 --> Looking for /usr/local/sbin/apxs: not found
 --> Looking for /usr/local/bin/apxs: not found
 --> Looking for /usr/sbin/apxs: not found
 --> Looking for /usr/bin/apxs: found
 --> Looking for /sbin/apxs: not found
 --> Looking for /bin/apxs: not found
 --> Looking for /usr/games/apxs: not found
 --> Looking for /usr/local/games/apxs: not found
 --> Looking for /snap/bin/apxs: not found
 --> Looking for /usr/local/rvm/bin/apxs: not found
 --> Looking for /home/ubuntu/.rvm/bin/apxs: not found
1 symlink duplicate(s) detected; ignoring them.

Analyzing /usr/bin/apxs...
Detecting main Apache executable...
 --> Looking for /usr/bin/apache2: not found
 --> Looking for /usr/sbin/apache2: found
Detecting version...
 --> 2.4.29
Detecting control command...
 --> Looking for /usr/bin/apache2ctl: not found
 --> Looking for /usr/bin/apachectl2: not found
 --> Looking for /usr/bin/apachectl: not found
 --> Looking for /usr/sbin/apache2ctl: found
Detecting configuration file location...
 --> /etc/apache2/apache2.conf
Detecting error log file...
 --> Attempting to substitute environment variables in Apache config value "${APACHE_LOG_DIR}/error.log"...
 --> Substituted "${APACHE_LOG_DIR}" -> "/var/log/apache2"
 --> /var/log/apache2/error.log
Detecting a2enmod and a2dismod...
 --> Looking for /usr/bin/a2enmod: not found
 --> Looking for /usr/sbin/a2enmod: found
 --> Looking for /usr/bin/a2dismod: not found
 --> Looking for /usr/sbin/a2dismod: found
Found a usable Apache installation using /usr/bin/apxs.

Final autodetection results
 * Found Apache 2.4.29!
   Information:
      apxs2          : /usr/bin/apxs
      Main executable: /usr/sbin/apache2
      Control command: /usr/sbin/apache2ctl
      Config file    : /etc/apache2/apache2.conf
      Error log file : /var/log/apache2/error.log

   To install Phusion Passenger against this specific Apache version:
      /usr/local/rvm/gems/ruby-2.4.9/wrappers/ruby /usr/bin/passenger-install-apache2-module --apxs2-path='/usr/bin/apxs'

   To start, stop or restart this specific Apache version:
      /usr/sbin/apache2ctl start
      /usr/sbin/apache2ctl stop
      /usr/sbin/apache2ctl restart

   To troubleshoot, please read the logs in this file:
      /var/log/apache2/error.log
```

```
$ passenger-config about root
/usr/lib/ruby/vendor_ruby/phusion_passenger/locations.ini
```

Relevant configuration section for `/etc/apache2/sites-enabled/default-ssl.conf`:

```
DocumentRoot /var/www/kaiapuni-test/public
    #***** Rails Environment settings
    RailsEnv development
    #***** Passenger Phusion environment settings
    # PassengerUserSwitching on
    # PassengerDefaultUser kaiapuni-test
    # Ruby Version to use with Passenger
    PassengerRuby /usr/local/rvm/gems/ruby-2.4.9/wrappers/ruby
    # Temp variable directory for Passenger to use for storing session information
    # PassengerInstanceRegistryDir /tmp
    # Hide passenger version information in HTTP header info
    PassengerShowVersionInHeader off
    # Enable high performance for Passenger to speed up functionality
    PassengerHighPerformance on
    # Set the maximum number of worker pools; needs to be higher than 6
    PassengerMaxPoolSize 6
    # Set a fixed number of running instances
    PassengerMinInstances 3
    # Pre-start the application for the given URL instead of waiting for the first HTTP request to do so
    PassengerPreStart https://kaiapuni-test.coe.hawaii.edu/
    # Max number of queued requests possible.
    PassengerMaxRequestQueueSize 250
```

OF NOTE:  I was testing manually adding `PassengerInstanceRegistryDir /tmp` but it did not seem to make much of a difference.

**Question 2: Passenger version and integration mode**:
- For example: open source 5.0.26 standalone; enterprise 5.0.21/nginx

Your answer:

Phusion Passenger Open Source 6.0.4 running with `mod_passenger` in Apache 2.4.29.

**Question 3: OS or Linux distro, platform (including version)**:
- For example: Debian 8, x86_64 or OS X 10.10 Yosemite, x86_64

Your answer:

Ubuntu 18.04

**Question 4: Passenger installation method**:

Your answer:
[ ] RubyGems + Gemfile
[ ] RubyGems, no Gemfile
[X] Phusion APT repo
[ ] Phusion YUM repo
[ ] OS X Homebrew
[ ] source tarball
[ ] Other, please specify:

Installed following instructions at https://www.phusionpassenger.com/library/install/apache/install/oss/bionic/

**Question 5: Your app's programming language (including any version managers) and framework (including versions)**:
- For example: Ruby 2.3.0, RVM, Rails 5.0.0; Node.js 4.0.0 with Express 4.13.4

Your answer:

Ruby 2.4.9 installed globally via RVM 1.2.9, Rails 4.2.11.1.

**Question 6: Are you using a PaaS and/or containerization? If so which one?**
- For example: Heroku, Amazon Container Services, Docker 1.9 with an image based on passenger-docker

Your answer:

No, self hosting, followed install instructions at https://www.phusionpassenger.com/library/install/apache/install/oss/bionic/

**Question 7: Anything else about your setup that we should know?**

Your answer: 

Passenger appears to be running and serving content without issue.  I have tried multiple times to restart Apache2 and have confirmed with `ps -aux` that it is happily chugging along and serving requests:

```
$ ps -aux | grep "Passenger"
root     24845  0.0  0.6 399456 13848 ?        Ssl  10:24   0:00 Passenger watchdog
root     24849  0.2  0.9 691668 19456 ?        Sl   10:24   0:08 Passenger core
kaiapun+ 25050  0.0  4.0 344240 82768 ?        Sl   10:24   0:00 Passenger AppPreloader: /var/www/kaiapuni-test (forking...)
kaiapun+ 25069  0.0  3.3 344340 68140 ?        Sl   10:24   0:00 Passenger AppPreloader: /var/www/kaiapuni-test (forking...)
kaiapun+ 25090  0.0  3.3 344440 68148 ?        Sl   10:24   0:00 Passenger AppPreloader: /var/www/kaiapuni-test (forking...)
ubuntu   28267  0.0  0.0  13136  1084 pts/0    S+   11:23   0:00 grep --color=auto Passenger
```

While investigating this issue, I ran `netstat` and noticed that there IS an active write open to the Passenger temp files as follows:

```
unix  3      [ ]         STREAM     CONNECTED     24303651 /tmp/passenger.57pfVS4/apps.s/ruby.ulQjNnS0WaYDaA7KoAcMHtkuYmeuhhqDkhDKPbbIx6YT9wrLgTrykb6KzxI0Hus
```

However, when I check that directory location, the file does not actually exist:

```
/tmp$ ls -al
total 60
drwxrwxrwt 15 root root   4096 Jan 28 11:20 .
drwxr-xr-x 24 root root   4096 Jan  8 06:21 ..
drwxrwxrwt  2 root root   4096 Jan  3 13:42 .font-unix
drwxrwxrwt  2 root root   4096 Jan  3 13:42 .ICE-unix
drwx------  2 root root   4096 Jan  3 13:47 netplan_a08ot19z
drwxr-x--x  2 root bareos 4096 Jan  7 02:00 sql
drwx------  3 root root   4096 Jan 28 10:24 systemd-private-4439ec0c8f174596bd1772cfe5841379-apache2.service-tuzxlf
drwx------  3 root root   4096 Jan 13 14:40 systemd-private-4439ec0c8f174596bd1772cfe5841379-memcached.service-RTDsfU
drwx------  3 root root   4096 Jan  3 13:42 systemd-private-4439ec0c8f174596bd1772cfe5841379-systemd-resolved.service-Z5nsWy
drwx------  3 root root   4096 Jan  7 12:35 systemd-private-4439ec0c8f174596bd1772cfe5841379-systemd-timesyncd.service-D5rulk
drwx------  3 root root   4096 Jan  3 13:42 systemd-private-4439ec0c8f174596bd1772cfe5841379-vnstat.service-o71s3M
drwxrwxrwt  2 root root   4096 Jan  3 13:42 .Test-unix
drwx------  2 root root   4096 Jan  3 13:42 vmware-root_627-4013264619
drwxrwxrwt  2 root root   4096 Jan  3 13:42 .X11-unix
drwxrwxrwt  2 root root   4096 Jan  3 13:42 .XIM-unix
```

I have NOT yet tried creating a different directory and pointing `PassengerInstanceRegistryrDir` at it.

I am able to run any necessary commands, updates, or changes to the server to assist with troubleshooting.

**Comment by CamJN** [comment]
I believe this is related to systemd launching apach using a private tmp dir. Can you check if that's the case?

**Comment by pkarjala** [comment]
Thank you; I checked as noted, and sure enough, that was the case.  I've disabled the `PrivateTmp` option in `/etc/systemd/system/multi-user.target.wants/apache2.service`.  Doing so allowed `passenger-status` to run without issue.

http://0pointer.de/blog/projects/security.html has more information on the repercussions of changing this configuration option in `systemd` should anyone run across this issue in the future.

Thank you for your assistance!
